packages/sdk/app-graph/src/graph-builder.ts

```diff
@@ -2,50 +2,236 @@
 // Copyright 2023 DXOS.org
 //
 
-import { EventSubscriptions, type UnsubscribeCallback } from '@dxos/async';
+import { type Signal, effect, signal } from '@preact/signals-core';
+// import { yieldOrContinue } from 'main-thread-scheduling';
+
+import { EventSubscriptions } from '@dxos/async';
+import { invariant } from '@dxos/invariant';
 
 import { Graph } from './graph';
+import { type EdgeDirection, type NodeArg, type Node } from './node';
+
+export type ConnectorExtension = (params: {
+  node: Node;
+  direction: EdgeDirection;
+  type?: string;
+}) => NodeArg<any>[] | undefined;
+export type HydratorExtension = (params: { id: string; type: string }) => NodeArg<any> | undefined;
+
+export type ExtensionType = 'connector' | 'hydrator';
+export type BuilderExtension =
+  | {
+      type: 'connector';
+      extension: ConnectorExtension;
+    }
+  | {
+      type: 'hydrator';
+      extension: HydratorExtension;
+    };
+
+export const connector = (extension: ConnectorExtension): BuilderExtension => ({ type: 'connector', extension });
```

Comment:
what is the purpose of this polymorhpism since these extensions do very different things?

Reply:
I don't think I'm hard set on them being the same thing, I think this mostly just evolved from convenience of not changing the api. One thing I'd say though is that I don't think they do very different things, they take different inputs but in the end they both add nodes to the graph.